### ISSUE TITLE
core_dist_n_jobs =1 or -1 -> no difference at all and computation time extremely high

### ISSUE THREAD

**Issue by matlabat** [scikit-learn-contrib/hdbscan#131]
Hi,
As the title say, I don't have any improvement by playing with the parameter core_dist_n_jobs.
I'm stuck at 18-20% on my CPU usage in any case.
Referring to the API Documentation, I've seen that the number of parallel jobs to run in core distance computations has to be **supported by the specific algorithm**.
In my case, it is set to "best".
So my first assumption is that the model has choosen an algorithm that can not be parallelized _given the nature of my data_

Is this assumption correct?
If yes:
How can I know which algorithm was actually selected?
What are the algorithm that can or cannot be parallelized in this task?

If I want to force an algorithm that can be parallelized and considering the nature of my data: 
Normalize(PCA(doc2vec on a Text corpus + additionnal features)) which give 
a 1 000 000 rows x 150 features matrix 
what would be the best shot according to you?

Also, if I subsample my dataset with:
10 000 rows -> the computation time is 24 sec
20 000 rows -> the computation time is 1 min 30 sec
30 000 rows -> the computation time is 3 min 24 sec
40 000 rows -> the computation time is 5 min 56 sec

we can see that the computation time seems to be **exponential in my case.**
If I roughly double the computation time every 10K rows, I could never treat 1M rows even in a night !

Do I have made an error with my other parameters that could blow this computation time?

HDBSCAN(algorithm='best', allow_single_cluster=False, alpha=1.0,
    approx_min_span_tree=True, cluster_selection_method='eom',
    core_dist_n_jobs=1, gen_min_span_tree=True, leaf_size=40,
    match_reference_implementation=False, memory=Memory(cachedir=None),
    metric='euclidean', min_cluster_size=20, min_samples=10, p=None,
    prediction_data=False)

Thanks for your help.
Mathieu

**Comment by lmcinnes** [comment]
Only a small part of the work can be easily parallelised (the core distance
computation). The next catch is that you have reasonably high dimensional
data which can significantly impact scaling performance. If you can do some
dimension reduction down to, say, 50 dimensions, you should see better
scaling performance. Sorry that we don't scale well for higher dimensional
data yet. we are working on it, but it is a non-trivial problem.

**Comment by matlabat** [comment]
Hi,
Thanks for your answer.
Unfortunately, I've already tried to decrease the dimension of my doc2vec model, but at a certain point it fails to catch all the specificity in my text corpus. As you can see, I already perform a dimension reduction with a PCA to fall at 150 feature vectors. 
I'm not an expert, but maybe I could ultimately try a non linear dimension reduction instead of PCA. However, I don't know much about good algorithms in my situation, if it will be performant to reduce considering that it comes from an NN (which is already "non linear"), and how to evaluate the variance kept in the new space. Any recommendation?
By default, I will turn myself with mini batch K-means which seems to be the only one to handle my problem in a decent time...
In any case, keep up the good work because it is very promising and I am/was very enthusiast to use your work.

Mathieu

**Comment by lmcinnes** [comment]
K-Means is sometimes the only option due to performance constraints. I can
offer [UMAP](https://github.com/lmcinnes/umap) as experimental non-linear
dimension reduction which should scale to reasonably large embedding
dimension (theoretically), but that's definitely in the experimental camp.
Worse it still uses KDTrees currently, so won't be as performant as it
could/should be. I have a version using NN-descent, but it isn't published
at this time. In short, I agree you may have to just stick with K-Means for
now; hopefully we can offer you better solutions in the near future.

**Comment by rtmatx** [comment]
Is there any plan to use `NN-descent` for K-NN graph approximation for HDBSCAN ? I have read the paper with a summary of it mentioned here [http://jacobj.ca/hdbscan.pdf](url). There is some performance gain for large data sets with some compromise on the results.

**Comment by lmcinnes** [comment]
I was actually working on using it in a new clustering algorithm that has a
lot of similarity to HDBSCAN, and where my current clustering effort is
right now. To add NN-descent to hdbscan would involve translating my
current numba implementation into Cython, or simply writing it a new
version fro the ground up. Neither is impossible, but either would be a
fairly large amount of work. If anyone is interested in taking this on I
would be happy to merge it, but I don't have the time to take on such a
project myself at this time.

Leland.

**Comment by lmcinnes** [comment]
The reference implementation is kgraph (https://github.com/aaalgo/kgraph),
but that's dense C++. I have a python/numba implementation (making use of
RP-trees as well) in PyNNDescent (https://github.com/lmcinnes/pynndescent).
The paper (http://www.cs.princeton.edu/cass/papers/www11.pdf) is pretty
readable and has useful pseudocode. Any progress you make is welcome, but I
am certainly understanding of the difficulties of finding time. Thanks for
even looking into this at all!

**Comment by rtmatx** [comment]
Thanks for all the links. I will start working on this during the weekends starting from tomorrow. Should I create a new issue to address any/all of the questions I may have during the implementation ?

**Comment by lmcinnes** [comment]
Yes, I think a new issue would be the best place for now. Thanks for
looking into this.
